### PR TITLE
Add capsule geometry use vtkCapsuleSource and vtkCylinderSource with `SetCapsuleCap`

### PR DESCRIPTION
**Pull request by kmarchais**
### Overview

Implementation of the capsule geometry following #5647 .
VTK 9.0 introduced `vtkCapsuleSource` which was deprecated in 9.3 and replaced by the hemispherical capping on the `vtkCylinderSource`.

I tried to implement this while taking care about the different versions but I am not sure how to manage this.
I also added corresponding tests.

Tested with VTK 9.3 and VTK 9.2.

| VTK 9.3 | 9.0 <= VTK <= 9.2 |
|----------|---------------------|
| image | image |

Edit: close #5647

### CODE REVIEW

**Comment by user27182** [comment]
Since the new `Capsule` object has a `direction`, the `direction_objects` tests will fail unless a new regression image is added for `Capsule`.

The regression tests were added very recently in #5614. However, since the images are generated in alphabetical order (but with numerical values`, it's likely that all direction object images which come after `Capsule` alphabetically will need to be renumbered. This means that many of the test images will show up as renamed and/or deleted. PR #5643 should fix this issue, such that only the new `direction_objects[Capsule-pos].png` and `direction_objects[Capsule-neg].png` will need to be added.

**Comment by user27182** [comment]
(Note: to generate the images locally you can run `pytest tests/plotting/ -k direction --generated_image_dir debug_images`, then copy the new images from `./debug_images` to `./tests/plotting/image_cache`)

**Comment by kmarchais** [comment]
Alright, thanks a lot for this @user27182 !
I've generated the images, copied and replaced the ones in `./tests/plotting/image_cache`. Now the test seems to pass, at least locally.
Should I commit this or wait for PR #5643 ?

**Comment by user27182** [comment]
@kmarchais no problem, yes please commit any changes to the image cache, it will be good to run all the CI tests to make sure everything else is still good.

**Comment by user27182** [comment]
Ok yes so as I thought it looks like pretty much all images had to be renumbered, which is not ideal since this PR otherwise should have no impact on the other images. I think #5643 should be merged first, but that will be up do the main devs to decide.

Also, it looks some other `direction` images were modifed in the cache, unrelated to the `direction_objects` test. These should probably be deleted / left unchanged in the image cache.

This command should be more specific and generate only the images from that test (`direction` is replaced with `direction_objects`):

`pytest tests/plotting/ -k direction_objects --generated_image_dir debug_images`

**Comment by kmarchais** [comment]
So it looks good with Python 3.12 which has VTK 9.3 but not with VTK < 9.3. 
I guess it's because of the difference between the capsules shown on the first message of this PR.

**Comment by user27182** [comment]
> So it looks good with Python 3.12 which has VTK 9.3 but not with VTK < 9.3.
> I guess it's because of the difference between the capsules shown on the first message of this PR.

One option may be to use `pytest.xfail`condtionally inside `test_direction_objects` if `name==Capusle` and `vtk_version < (9, 3)`. Perhaps better would be to increase the regression error tolerance conditionally just for that test, though I'm not sure if that's possible.

If the geometries are substantially different across VTK versions, is it worth adding a warning to the docs, to warn users about this?

**Comment by kmarchais** [comment]
Here is the difference caused by the way the resolution is managed in both versions.

image

In VTK 9.3, the resolution follows the one of a Cylinder while before 9.3 it has a `phi_resolution` and a `theta_resolution` like a Sphere.

In the new `Capsule` function in `pyvista/core/utilities/geometric_objects.py` both `phi` and `theta` resolutions are given but only `theta_resolution` is used as `resolution` in the VTK 9.3 case.
To follow VTK's structure, maybe it would be better to have the `Capsule` function only for 9.0 <= VTK <= 9.2 since the capsule is managed with the `vtkCylinderSource` from 9.3.
But I am not sure how to manage this correctly, especially with regression tests.

**Comment by user27182** [comment]
> In the new `Capsule` function in `pyvista/core/utilities/geometric_objects.py` both `phi` and `theta` resolutions are given but only `theta_resolution` is used as `resolution` in the VTK 9.3 case. 

I'm not familiar with the use cases of `Capsule` objects, but would it be acceptable in this case to just have a single `resolution` parameter, which internally would set `phi_resolution = theta_resolution = resolution` for VTK < 9.3 ?

> To follow VTK's structure, maybe it would be better to have the `Capsule` function only for 9.0 <= VTK <= 9.2 since the capsule is managed with the `vtkCylinderSource` from 9.3. 

If capsules have a distinct use cases, I think having a unified `Capsule` API is good, regardless of `vtk` version. Even if it uses `vtkCylinderSource` for 9.3,  having rounded hemispheres by default would at least make `Capsule` a convenience/wrapper function.

> But I am not sure how to manage this correctly, especially with regression tests.

Given that the geometries do differ substantially in terms of their mesh construction, we should probably test these separately. I think adding a conditional `pytest.xfail` in `test_direction_objects` for the < 9.3 cases should be OK. We can test the deprecated < 9.3 case separately.

**Comment by kmarchais** [comment]
Trying with `verify_image_cache.skip = name == 'Capsule' and pv.vtk_version_info < (9, 3, 0)` since `verify_image_cache` is used in other tests.
Is this similar to using `xfail` @user27182 ?

File: tests/plotting/test_plotting.py
Comment (user27182) [inline tests/plotting/test_plotting.py:4119]:
```suggestion

    # Test Capsule separately based on vtk version
    if 'Capsule' in name:
       legacy_vtk = pv.vtk_version_info.major < 9 and pv.vtk_version_info.minor < 3
       if legacy_vtk and 'legacy' not in name or not legacy_vtk and 'legacy' in name:
            pytest.xfail('Test capsule separately for different vtk versions. Expected to fail if testing with wrong version.')
            
```

File: tests/plotting/test_plotting.py
Comment (user27182) [inline tests/plotting/test_plotting.py:4117]:
```suggestion
def test_direction_objects(object_function, positive_dir):
```

File: tests/plotting/test_plotting.py
Comment (user27182) [inline tests/plotting/test_plotting.py:4127]:
```suggestion

```

File: tests/plotting/test_plotting.py
Comment (user27182) [inline tests/plotting/test_plotting.py:4111]:
```suggestion

        # Add a separate test for vtk < 9.3
        expected_names.append('Capsule_legacy')
        actual_names.append('Capsule_legacy')
    
    assert sorted(actual_names) == sorted(expected_names)
```

**Review by user27182 — COMMENTED** [review]
Here is an alternative where an extra test case is added. This way, regression images are generated for both new and legacy versions of Capsule. No skips are used, instead the tests will still run for mis-matched vtk versions but are marked as expected to fail.

**Comment by user27182** [comment]
> Trying with `verify_image_cache.skip = name == 'Capsule' and pv.vtk_version_info < (9, 3, 0)` since `verify_image_cache` is used in other tests.
> Is this similar to using `xfail` @user27182 ?

This would have similar effect yes. I think there are subtle differences, such as skipping altogether means the test isn't executed, whereas `xfail` will still run the test. Also, using `xfail` would show up in test summaries reported by pytest, wherase `verify_image_cache.skip` may not, though this may not be a problem.

**Comment by user27182** [comment]
I think the main thing is that we may not actually want to completely skip the tests for the legacy Capsule, so we can try to make a separate regression test altogether just for that case, or manually add a new case to the `test_direction_objects` test

**Comment by user27182** [comment]
> image

With the additional test cases, tests with vtk version < 9.3 will test against the legacy image. So, two regression images (positive and negative direction) need to be added for both cases shown in these screenshots.

Ideally, with #5643, these would be:

- `direction_objects[Capsule-pos].png`
- `direction_objects[Capsule-neg].png`
- `direction_objects[Capsule_legacy-pos].png`
- `direction_objects[Capsule_legacy-neg].png`

**Comment by kmarchais** [comment]
Ok great! 
With the current naming is there a way to make it pass?
I think it will be named `object_function1` in both cases.

**Comment by user27182** [comment]
> I think it will be named `object_function1` in both cases.

These lines should be generating both a `Capsule` and `Capsule_legacy` test case, regardless of vtk version. So in theory `Capsule` -> `object_function1` and `Capsule_legacy` -> `object_function2`... but it seems like this isn't the case?https://github.com/pyvista/pyvista/blob/64b409a0db6ed1eda636730546a75ad18ec1a221/tests/plotting/test_plotting.py#L4104-L4114

I would expect that the generated images with vtk 9.3 would result in images for `object_function1` and `object_function2` being identical. Running again with vtk < 9.3 would generate the legacy images, but again `object_function1` and `object_function2` would be identical. 

And then copy the 9.3 version images for `object_function1` to the cache, and copy the <9.3 images for `object_function2` to the cache. After copying, running `pytest` locally  vtk 9.3 the should result in `object_function1` tests passing, and `object_function2` tests failing. But running the tests again with vtk <9.3 should result in the opposite cases passing/failing. 

Also, I would also expect that all downstream images would (again) have to be renumbered and re-commited.

File: tests/plotting/test_plotting.py
Comment (user27182) [inline tests/plotting/test_plotting.py:4115]:
```suggestion

```

File: tests/plotting/test_plotting.py
Comment (user27182) [inline tests/plotting/test_plotting.py:4104]:
```suggestion

```

File: tests/plotting/test_plotting.py
Comment (user27182) [inline tests/plotting/test_plotting.py:4072]:
```suggestion
        'Capsule',
        'Capsule_legacy',
        'CircularArcFromNormal',
```

**Review by user27182 — COMMENTED** [review]
The issue was that the legacy case was not added to the actual test case dictionary.

Also, the conditional for pyvista 0.43 can be safely removed (it was added only for testing part of #5435 )

File: tests/plotting/test_plotting.py
Comment (user27182) [inline tests/plotting/test_plotting.py:4068]: This is the key line needed to add the new test case
```suggestion
functions['Capsule_legacy'] = functions['Capsule']
```

**Review by MatthewFlamm — COMMENTED** [review]
I left some comments for the PR as is.  I haven't fully thought this one through, but would it better to simply have this option only be available for vtk >9.3 and expose this option in Cylinder as a parameter? This would reduce maintenance burden, but reduce availability for vtk <9.3.  Given 9.3 is just released, I'd probably lean towards the current PR state.

I think we should wait for the changes to the test images for geometric objects with direction for this one.

File: pyvista/core/utilities/geometric_objects.py
Comment (kmarchais) [reply pyvista/core/utilities/geometric_objects.py:110]: Thanks!

File: tests/plotting/test_plotting.py
Comment (user27182) [inline tests/plotting/test_plotting.py:4103]:
```suggestion
        'SolidSphere', 
        'SolidSphereGeneric',
        'Sphere',
        'Text3D'
```

I should have re-added these here when I suggested removing them from the conditional 0.43 block

Thanks for your patience dealing with these changes!

File: pyvista/core/utilities/geometric_sources.py
Comment (kmarchais) [reply pyvista/core/utilities/geometric_sources.py]: This seems to be causing a circular import 

```
if pyvista.vtk_version_info <= (9, 2):
AttributeError: partially initialized module 'pyvista' has no attribute 'vtk_version_info' (most likely due to a circular import)
```

**Comment by user27182** [comment]
Looks like #5643 will be approved, sorry for all the trouble @kmarchais dealing with the test images. Once that PR is merged the image cache here should be easier to work with, and only the 4 named images mentioned earlier will need to be added.

**Comment by kmarchais** [comment]
Ok perfect! Thanks for the help!

**Comment by kmarchais** [comment]
> > I think it will be named `object_function1` in both cases.
> 
> These lines should be generating both a `Capsule` and `Capsule_legacy` test case, regardless of vtk version. So in theory `Capsule` -> `object_function1` and `Capsule_legacy` -> `object_function2`... but it seems like this isn't the case?
> 
> https://github.com/pyvista/pyvista/blob/64b409a0db6ed1eda636730546a75ad18ec1a221/tests/plotting/test_plotting.py#L4104-L4114
> 
> I would expect that the generated images with vtk 9.3 would result in images for `object_function1` and `object_function2` being identical. Running again with vtk < 9.3 would generate the legacy images, but again `object_function1` and `object_function2` would be identical.
> 
> And then copy the 9.3 version images for `object_function1` to the cache, and copy the <9.3 images for `object_function2` to the cache. After copying, running `pytest` locally vtk 9.3 the should result in `object_function1` tests passing, and `object_function2` tests failing. But running the tests again with vtk <9.3 should result in the opposite cases passing/failing. All tests should now pass locally for any vtk version since the conditional should correctly mark if a test is expected to fail
> 
> Also, I would also expect that all downstream images would (again) have to be renumbered and re-commited.

I keep having only one version generated (`function1`), `function2` being `CircularArcFromNormal`. I may be missing something though.

**Comment by user27182** [comment]
> I keep having only one version generated (`function1`), `function2` being `CircularArcFromNormal`. I may be missing something though.

The `Capsule_legacy` case is added to the end of the test cases dict. And the dict itself isn't sorted before the test cases are processed by pytest. So, I think `Capsule_legacy` may actually be `object_function32` or something like that instead of `object_function2`. So, you could sort the dict as part of the `return` statement, or look for item 32 in the image cache, or wait until #5643 is merged, which should fix all these issues with the numbered test cases.

File: pyvista/core/utilities/geometric_sources.py
Comment (MatthewFlamm) [inline pyvista/core/utilities/geometric_sources.py:64]: A different suggestion to avoid the circular import.  We only support vtk >=9.

```suggestion
if _vtk.vtkVersion.GetVTKMajorVersion() == 9 and _vtk.vtkVersion.GetVTKMinorVersion() <= 2:
```

Comment (MatthewFlamm) [reply pyvista/core/utilities/geometric_sources.py:64]: Hmmm.  I wonder if we need to move `vtk_version_info` instead.  I'd suggest inside `_vtk_core`.  But that isn't for this PR.  Let another suggestion.

**Comment by kmarchais** [comment]
I have just added `Capsule` and `Capsule_legacy` to `image_cache` but I also had to add all the `direction_objects` images to pass the tests locally.
Also, running the command `pytest tests/plotting/ -k direction_objects --generated_image_dir debug_images` generated 69 `direction_objects[Capsule-pos]_*.png` extra files (70 being the total number of `direction_objects` images)  and I don(t know why.

**Comment by user27182** [comment]
> I have just added `Capsule` and `Capsule_legacy` to `image_cache` but I also had to add all the `direction_objects` images to pass the tests locally.

You can try running 
`git checkout --patch main tests/plotting/image_cache`

This should replace all images in your cache locally with the original ones from main for all direction objects. Do the tests still fail locally then?

> Also, running the command `pytest tests/plotting/ -k direction_objects --generated_image_dir debug_images` generated 69 `direction_objects[Capsule-pos]_*.png` extra files (70 being the total number of `direction_objects` images)  and I don't know why.

When you run the plotting tests and there is a missing image in the cache, pytest will sometimes re-try the test automatically. This can mean it will generate many images in `debug_images` (one for each test attempt). The solution is to copy just one of these images to `image_cache`. The next time you run the test it should be OK since now there is no missing image and pytest won't retry the test.

**Comment by kmarchais** [comment]
> You can try running 
> `git checkout --patch main tests/plotting/image_cache`
This tells me `cannot apply binary patch` ans `patch does not apply`.

When looking at the differences between images using swipe mode, the positions of some axes labels differ.

**Comment by user27182** [comment]
> > You can try running 
> 
> > `git checkout --patch main tests/plotting/image_cache`
> 
> 
> This tells me `cannot apply binary patch` ans `patch does not apply`.

You can try deleting all images in your local image_cache first (except for the 4 new capsule images), and then try the command again.

> When looking at the differences between images using swipe mode, the positions of some axes labels differ.

I have not reviewed the diff between main and this branch for the direction object images, but sometimes there are minor differences between images generated on Windows vs MacOS vs Linux. There is a small tolerance in the tests to allow for this.

You can try committing the original images and push the changes anyway, see how the CI tests do

**Comment by kmarchais** [comment]
> You can try deleting all images in your local image_cache first (except for the 4 new capsule images), and then try the command again.

I tried this also but didn't change. I probably don't do it the right way.
So I put back the original images and I think it is good now.

File: pyvista/core/utilities/__init__.py
Comment (tkoyama010) [inline pyvista/core/utilities/__init__.py:107]:
```suggestion
except ImportError:  # pragma: no cover
```

File: pyvista/core/utilities/geometric_objects.py
Comment (tkoyama010) [inline pyvista/core/utilities/geometric_objects.py:50]:
```suggestion
except ImportError:  # pragma: no cover
```

File: pyvista/core/utilities/geometric_objects.py
Comment (tkoyama010) [inline pyvista/core/utilities/geometric_objects.py:111]:
```suggestion
    if pyvista.vtk_version_info >= (9, 3):  # pragma: no cover
```

File: tests/core/test_geometric_sources.py
Comment (tkoyama010) [inline tests/core/test_geometric_sources.py:58]:
```suggestion
@pytest.mark.needs_vtk_version(9, 3, 0)
def test_cylinder_capsule_cap():
    algo = pv.CylinderSource()
    algo.capsule_cap = True
    assert algo.capsule_cap
```

File: tests/core/test_geometric_sources.py
Comment (tkoyama010) [inline tests/core/test_geometric_sources.py:21]:
```suggestion
@pytest.mark.needs_vtk_version(9, 3, 0)
def test_capsule_source():
    algo = pv.CapsuleSource()
    assert np.array_equal(algo.center, (0.0, 0.0, 0.0))
    assert np.array_equal(algo.direction, (1.0, 0.0, 0.0))
    assert algo.radius == 0.5
    assert algo.cylinder_length == 1.0
    assert algo.theta_resolution == 30
    assert algo.phi_resolution == 30
    direction = np.random.default_rng().random(3)
    algo.direction = direction
    assert np.array_equal(algo.direction, direction)
```

File: tests/core/test_geometric_sources.py
Comment (tkoyama010) [inline tests/core/test_geometric_sources.py:414]:
```suggestion
```

File: tests/plotting/test_plotting.py
Comment (tkoyama010) [inline tests/plotting/test_plotting.py:4150]:
```suggestion
        legacy_vtk = pv.vtk_version_info < (9, 3)
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
Awesome job!

**Comment by kmarchais** [comment]
Thanks a lot for all the help!

**Comment by tkoyama010** [comment]
github-actions preview

**Review by tkoyama010 — APPROVED** [review]
LGTM
<p align="center">

</p>

**Comment by tkoyama010** [comment]
@namheegordonkim Could you confirm that this will also meet your needs in #5647?

File: pyvista/core/utilities/geometric_sources.py
Comment (user27182) [inline pyvista/core/utilities/geometric_sources.py:72]:
```suggestion

.. versionadded:: 0.44.0

```

File: pyvista/core/utilities/geometric_sources.py
Comment (user27182) [inline pyvista/core/utilities/geometric_sources.py:731]:
```suggestion

.. versionadded:: 0.44.0

```

**Review by user27182 — COMMENTED** [review]
LGTM, some minor suggestions for docstrings

**Comment by tkoyama010** [comment]
pre-commit.ci autofix

**Comment by tkoyama010** [comment]
pre-commit.ci autofix

**Review by tkoyama010 — APPROVED** [review]

**Comment by namheegordonkim** [comment]
Haven't pulled these to check with the rest of my visualizer code, but from the looks of it, it definitely fits my bill! Thank you for the awesome work.

**Pull request merged**
